collectors/perf.cpp

```diff
@@ -79,14 +74,33 @@ static inline uint64_t makeup_booker_ci_config(int nodetype, int eventid, int by
     return config;
 }
 
-PerfCollector::PerfCollector(const Json::Value& config, const std::string& name) : Collector(config, name)
+PerfCollector::PerfCollector(const Json::Value& config, const std::string& name, bool enablePerapiPerf) : Collector(config, name), mEnablePerapiPerf(enablePerapiPerf)
 {
-    struct event leader = {"CPUCycleCount", PERF_TYPE_HARDWARE, PERF_COUNT_HW_CPU_CYCLES};
 
+    struct event leader;
+    if (mEnablePerapiPerf)
+    {
+        volatile uint64_t pmcr_el0;
+        asm volatile("mrs %0, PMCR_EL0" : "=r"(pmcr_el0));
+        pmu_counter_bits = ((pmcr_el0 & 0x80) == 0x80 ? 64 : 32);
+        if (!mConfig.isMember("type")) exit(EXIT_FAILURE);
```

Comment:
is there a "type" field in mConfig?

Reply:
Yes.  I found this type is different for different platforms. And if the type is wrong, PMEVCNTR2_EL0 register will read nothing. So I wonder if it's possible to set by json file.